### PR TITLE
Enable the load extension API

### PR DESCRIPTION
**Pull request by ksshannon**
Allow enabling the load extension mechanism and actually loading extensions.  This enables the C API only, not the SQL API.

### CODE REVIEW

**Comment by ksshannon** [comment]
Whoops, All snuck in somewhere, I'll send a new request.

**Comment by ksshannon** [comment]
Sorry for the messy PR...

File: extension_test.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	"runtime"
+	"strings"
+	"testing"
+
+	"crawshaw.io/sqlite"
+)
+
+const (
+	extCode = `
+/*
```
Comment (crawshaw) [inline extension_test.go]: you can skip the copyright notice in a temporary file.

File: extension_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#include "sqlite3ext.h"
+SQLITE_EXTENSION_INIT1
+
+#include <stdlib.h>
+
+static void hellofunc(
+	sqlite3_context *context,
+	int argc,
+	sqlite3_value **argv){
+		(void)argc;
```
Comment (crawshaw) [inline extension_test.go]: interesting style. is there a particular linter or guide you are working to?

File: extension_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		t.Skipf("no gcc support: %s, %s", string(out), err)
+	}
+	c, err := sqlite.OpenConn(":memory:", 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		err := c.Close()
+		if err != nil {
+			t.Error(nil)
```
Comment (crawshaw) [inline extension_test.go]: t.Error(err)

File: extension_test.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+	fout, err := os.Create(filepath.Join(tmpdir, "ext.c"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(fout, strings.NewReader(extCode))
+	fout.Close()
```
Comment (crawshaw) [inline extension_test.go]: as you are writing a file, check the error here and on the copy.

as that gets tedious, use ioutile:

if err := ioutile.WriteFile(file path.Join(tmpdir, "ext.c")), []byte(extCode), 0666); err != nil {
...
}

File: extension_test.go (hunk 1 above)
Comment (ksshannon) [reply extension_test.go]: Yes, of course...

File: extension_test.go (hunk 2 above)
Comment (ksshannon) [reply extension_test.go]: No, I started from the example on https://www.sqlite.org/loadext.html#write, and went from there.  SQLite's style is similar, I substituted tabs for spaces.

File: extension_test.go (hunk 3 above)
Comment (ksshannon) [reply extension_test.go]: whoops...

File: extension_test.go (hunk 4 above)
Comment (ksshannon) [reply extension_test.go]: Understood.

**Comment by ksshannon** [comment]
Thanks for the feedback, I'll make the final changes when I get a chance.

**Comment by ksshannon** [comment]
As a side note, I added this so I could load the spatialite extension, but it seems Dr. Hipp is added base functionality as a new extension:

https://sqlite.org/src/timeline?r=rtree-geopoly&nd&c=2018-08-23+14%3A54%3A45&n=200

I'm always amazed at the quality of SQLite.

**Pull request merged**